### PR TITLE
Refactor class structure to support Asynchronous implementation

### PR DESCRIPTION
**Pull request by justing-bq**
This PR reworks the class structure and to allow for multiple implementations and stubs out the asynchronous functionality.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: Amazon.QLDB.Driver/driver/BaseQldbDriver.cs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    /// <summary>
+    /// Base class for QLDB Drivers.
+    /// </summary>
+    public abstract class BaseQldbDriver
+    {
+        internal const string TableNameQuery =
+            "SELECT VALUE name FROM information_schema.user_tables WHERE status = 'ACTIVE'";
+
+        protected static readonly RetryPolicy DefaultRetryPolicy = RetryPolicy.Builder().Build();
+        internal readonly SessionPool sessionPool;
```
Comment (simonz-bq) [inline Amazon.QLDB.Driver/driver/BaseQldbDriver.cs:25]: All of these can be private protected I believe.

File: Amazon.QLDB.Driver/driver/AsyncQldbDriver.cs [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+ */
+
+namespace Amazon.QLDB.Driver
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using Amazon.Runtime;
+
+    /// <summary>
```
Comment (simonz-bq) [inline Amazon.QLDB.Driver/driver/AsyncQldbDriver.cs:21]: Please strip all docstrings + implementation. They shouldnt be there to begin with, and it will make reading future PRs impossible and a pain in the butt.

File: Amazon.QLDB.Driver/executable/IAsyncExecutable.cs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    {
+        /// <summary>
+        /// Execute the statement against QLDB and retrieve the result.
+        /// </summary>
+        ///
+        /// <param name="statement">The PartiQL statement to be executed against QLDB.</param>
+        ///
+        /// <returns>Result from executed statement.</returns>
+        /// <exception cref="AmazonServiceException">Thrown when there is an error executing against QLDB.</exception>
+        IAsyncResult Execute(string statement);
```
Comment (simonz-bq) [inline Amazon.QLDB.Driver/executable/IAsyncExecutable.cs:33]: The signatures for these are all incorrect.

File: Amazon.QLDB.Driver/driver/AsyncQldbDriver.cs [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+        /// and end the session.
+        /// </summary>
+        ///
+        /// <param name="action">The Executor lambda with no return value representing the block of code to be executed within the transaction.
+        /// This cannot have any side effects as it may be invoked multiple times.</param>
+        ///
+        /// <exception cref="TransactionAbortedException">Thrown if the Executor lambda calls <see cref="AsyncTransactionExecutor.Abort"/>.</exception>
+        /// <exception cref="QldbDriverException">Thrown when called on a disposed instance.</exception>
+        /// <exception cref="AmazonServiceException">Thrown when there is an error executing against QLDB.</exception>
+        public void Execute(Action<AsyncTransactionExecutor> action)
```
[the remark this answers is not in the document]
Comment (justing-bq) [reply Amazon.QLDB.Driver/driver/AsyncQldbDriver.cs:65]: That stuff will be part of the async work that comes later.

File: Amazon.QLDB.Driver/executable/IAsyncExecutable.cs (hunk 3 above)
Comment (justing-bq) [reply Amazon.QLDB.Driver/executable/IAsyncExecutable.cs:33]: These are not intended to be the final signatures.

File: Amazon.QLDB.Driver/driver/AsyncQldbDriver.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (justing-bq) [reply Amazon.QLDB.Driver/driver/AsyncQldbDriver.cs:65]: Also....you can't use `void` as a type argument.

File: Amazon.QLDB.Driver/driver/AsyncQldbDriver.cs [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+        /// <param name="action">The Executor lambda with no return value representing the block of code to be executed within the transaction.
+        /// This cannot have any side effects as it may be invoked multiple times.</param>
+        /// <param name="retryAction">A lambda that is invoked when the Executor lambda is about to be retried due to
+        /// a retriable error. Can be null if not applicable.</param>
+        ///
+        /// <exception cref="TransactionAbortedException">Thrown if the Executor lambda calls <see cref="AsyncTransactionExecutor.Abort"/>.</exception>
+        /// <exception cref="QldbDriverException">Thrown when called on a disposed instance.</exception>
+        /// <exception cref="AmazonServiceException">Thrown when there is an error executing against QLDB.</exception
+        [Obsolete("As of release 1.0, replaced by 'retryPolicy'. Will be removed in the next major release.")]
+        public void Execute(Action<AsyncTransactionExecutor> action, Action<int> retryAction)
```
[the remark this answers is not in the document]
Comment (justing-bq) [reply Amazon.QLDB.Driver/driver/AsyncQldbDriver.cs]: Removed.

File: Amazon.QLDB.Driver/driver/BaseQldbDriver.cs (hunk 1 above)
Comment (justing-bq) [reply Amazon.QLDB.Driver/driver/BaseQldbDriver.cs]: Done.

File: Amazon.QLDB.Driver/driver/AsyncQldbDriver.cs (hunk 2 above)
Comment (justing-bq) [reply Amazon.QLDB.Driver/driver/AsyncQldbDriver.cs]: Removed docstrings for async methods/classes (but left them for methods I don't see changing).

File: Amazon.QLDB.Driver/result/IAsyncResult.cs [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+namespace Amazon.QLDB.Driver
+{
+    using System.Collections.Generic;
+    using Amazon.IonDotnet.Tree;
+
+    /// <summary>
+    /// Interface for the result of executing a statement in QLDB.
+    /// Implements IEnumerable to allow iteration over Ion values within the result.
+    /// </summary>
+    public interface IAsyncResult : IEnumerable<IIonValue>
```
Comment (simonz-bq) [inline Amazon.QLDB.Driver/result/IAsyncResult.cs:23]: IAsyncEnumerable

Please update the rest of the result classes to account for this.

File: Amazon.QLDB.Driver/result/AsyncBufferedResult.cs [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        private readonly TimingInformation? timingInformation;
+
+        /// <summary>
+        /// Prevents a default instance of the <see cref="AsyncBufferedResult"/> class from being created.
+        /// </summary>
+        ///
+        /// <param name="values">Buffer values.</param>
+        /// <param name="consumedIOs">IOUsage statistics.</param>
+        /// <param name="timingInformation">TimingInformation statistics.</param>
+        private AsyncBufferedResult(List<IIonValue> values, IOUsage? consumedIOs, TimingInformation? timingInformation)
```
Comment (simonz-bq) [inline Amazon.QLDB.Driver/result/AsyncBufferedResult.cs:38]: Please remove the implementation here. We still don't know what it will be like just yet.

File: Amazon.QLDB.Driver/result/AsyncResult.cs [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+        private readonly IonEnumerator ionEnumerator;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AsyncResult"/> class.
+        /// </summary>
+        ///
+        /// <param name="session">The parent session that represents the communication channel to QLDB.</param>
+        /// <param name="statementResult">The result of the statement execution.</param>
+        /// <param name="txnId">The unique ID of the transaction.</param>
+        internal AsyncResult(Session session, string txnId, ExecuteStatementResult statementResult)
```
Comment (simonz-bq) [inline Amazon.QLDB.Driver/result/AsyncResult.cs:30]: Please remove the implementation this file as well. These don't seem right.

File: Amazon.QLDB.Driver/result/BaseBufferedResult.cs [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    {
+        protected readonly List<IIonValue> values;
+        protected readonly IOUsage? consumedIOs;
+        protected readonly TimingInformation? timingInformation;
+
+        /// <summary>
+        /// Returns an enumerator that iterates through a collection.
+        /// </summary>
+        /// <returns>An <see cref="IEnumerator"/> object that can be used to iterate through the collection.</returns>
+        public IEnumerator<IIonValue> GetEnumerator()
```
Comment (simonz-bq) [inline Amazon.QLDB.Driver/result/BaseBufferedResult.cs:35]: I do not believe this class would live in the base. The enumerator type is not sharable to the async impl.

File: Amazon.QLDB.Driver/result/BaseResult.cs [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+        public TimingInformation? GetTimingInformation()
+        {
+            return this.ionEnumerator.GetTimingInformation();
+        }
+
+        /// <summary>
+        /// Object which allows for iteration over the individual Ion values that make up the whole result of a statement
+        /// execution against QLDB.
+        /// </summary>
+        protected class IonEnumerator : IEnumerator<IIonValue>
```
Comment (simonz-bq) [inline Amazon.QLDB.Driver/result/BaseResult.cs:72]: This entire class is not sharable to the async implementation.

File: Amazon.QLDB.Driver/retry/AsyncRetryHandler.cs [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AsyncRetryHandler"/> class.
+        /// </summary>
+        /// <param name="logger">The logger to record retries.</param>
+        public AsyncRetryHandler(ILogger logger)
+        {
+            this.logger = logger;
+        }
+
+        public Task<T> RetriableExecute<T>(Func<T> func, RetryPolicy retryPolicy, Action newSessionAction, Action nextSessionAction, Action<int> retryAction)
```
Comment (simonz-bq) [inline Amazon.QLDB.Driver/retry/AsyncRetryHandler.cs:33]: I believe it's Func<Task<T>>

File: Amazon.QLDB.Driver/retry/IAsyncRetryHandler.cs [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+ */
+
+namespace Amazon.QLDB.Driver
+{
+    using System;
+    using System.Threading.Tasks;
+
+    internal interface IAsyncRetryHandler
+    {
+        Task<T> RetriableExecute<T>(Func<T> func, RetryPolicy retryPolicy, Action newSessionAction, Action nextSessionAction, Action<int> retryAction);
```
Comment (simonz-bq) [inline Amazon.QLDB.Driver/retry/IAsyncRetryHandler.cs:21]: Func<Task<T>>

File: Amazon.QLDB.Driver/session/AsyncQldbSession.cs [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+        /// <summary>
+        /// Create a transaction object which allows for granular control over when a transaction is aborted or committed.
+        /// </summary>
+        ///
+        /// <returns>The newly created transaction object.</returns>
+        public virtual IAsyncTransaction StartTransaction()
+        {
+            try
+            {
+                var startTransactionResult = this.session.StartTransaction();
```
Comment (simonz-bq) [inline Amazon.QLDB.Driver/session/AsyncQldbSession.cs:59]: Please remove the implementation. This is clearly incorrect.

File: Amazon.QLDB.Driver/session/AsyncSessionPool.cs [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+        public AsyncSessionPool(Func<Session> sessionCreator, IAsyncRetryHandler retryHandler, int maxConcurrentTransactions, ILogger logger)
+        {
+            this.sessionPool = new BlockingCollection<AsyncQldbSession>(maxConcurrentTransactions);
+            this.poolPermits = new SemaphoreSlim(maxConcurrentTransactions, maxConcurrentTransactions);
+            this.sessionCreator = sessionCreator;
+            this.retryHandler = retryHandler;
+            this.logger = logger;
+        }
+
+        public Task<T> Execute<T>(Func<AsyncTransactionExecutor, T> func, RetryPolicy retryPolicy, Action<int> retryAction)
```
Comment (simonz-bq) [inline Amazon.QLDB.Driver/session/AsyncSessionPool.cs:46]: Func<AsyncTransactionExecutor, Task<T>>

File: Amazon.QLDB.Driver/session/AsyncSessionPool.cs [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+                this.sessionPool.Add(session);
+            }
+
+            this.poolPermits.Release();
+            this.logger.LogDebug("Session returned to pool; pool size is now {}.", this.sessionPool.Count);
+        }
+
+        private AsyncQldbSession StartNewSession()
+        {
+            return new AsyncQldbSession(this.sessionCreator.Invoke(), this.ReleaseSession, this.logger);
```
Comment (simonz-bq) [inline Amazon.QLDB.Driver/session/AsyncSessionPool.cs:69]: This makes me question sessionCreator and Invoke. It seems like that would be async, and this does not account for that.

File: Amazon.QLDB.Driver/session/SessionPool.cs [hunk 16]
```diff
@@ -82,24 +80,12 @@ public T Execute<T>(Func<TransactionExecutor, T> func, RetryPolicy retryPolicy,
...
-            this.isClosed = true;
-            while (this.sessionPool.Count > 0)
-            {
-                this.sessionPool.Take().Close();
-            }
-        }
-
-        /// <summary>
-        /// <para>Get a <see cref="IQldbSession"/> object.</para>
+        /// <para>Get a <see cref="QldbSession"/> object.</para>
```
Comment (simonz-bq) [inline Amazon.QLDB.Driver/session/SessionPool.cs:83]: Why was this changed?

**Review by simonz-bq — CHANGES_REQUESTED** [review]

File: Amazon.QLDB.Driver/driver/AsyncQldbDriverBuilder.cs [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+            this.logRetries = true;
+            return this;
+        }
+
+        /// <summary>
+        /// Build a driver instance using the current configuration set with the builder.
+        /// </summary>
+        ///
+        /// <returns>A newly created driver.</returns>
+        public AsyncQldbDriver Build()
```
[the remark this answers is not in the document]
Comment (justing-bq) [reply Amazon.QLDB.Driver/driver/AsyncQldbDriverBuilder.cs:121]: The sync driver does this as well.

File: Amazon.QLDB.Driver/driver/BaseQldbDriver.cs [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+ */
+
+namespace Amazon.QLDB.Driver
+{
+    /// <summary>
+    /// Base class for QLDB Drivers.
+    /// </summary>
+    public abstract class BaseQldbDriver
+    {
+        internal const string TableNameQuery =
```
[the remark this answers is not in the document]
Comment (justing-bq) [reply Amazon.QLDB.Driver/driver/BaseQldbDriver.cs:21]: `private protected` results in the test code not compiling.

File: Amazon.QLDB.Driver/session/SessionPool.cs (hunk 16 above)
Comment (justing-bq) [reply Amazon.QLDB.Driver/session/SessionPool.cs:83]: Because IQldbSession does not exist in the code.

File: Amazon.QLDB.Driver/driver/IAsyncQldbDriver.cs [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+    using System.Collections.Generic;
+    using System.Threading.Tasks;
+
+    public interface IAsyncQldbDriver : IDisposable
+    {
+        Task Execute(Action<AsyncTransactionExecutor> action);
+
+        Task Execute(Action<AsyncTransactionExecutor> action, RetryPolicy retryPolicy);
+
+        Task<T> Execute<T>(Func<AsyncTransactionExecutor, T> func);
```
[the remark this answers is not in the document]
Comment (justing-bq) [reply Amazon.QLDB.Driver/driver/IAsyncQldbDriver.cs]: Done.

File: Amazon.QLDB.Driver/driver/BaseQldbDriverBuilder.cs [hunk 20]
```diff
@@ -0,0 +1,101 @@
...
+        /// Gets or sets the logger to create a driver with.
+        /// </summary>
+        private protected ILogger Logger { get; set; } = NullLogger.Instance;
+
+        /// <summary>
+        /// Gets or sets the configuration to construct the <see cref="AmazonQLDBSessionClient"/> object.
+        /// </summary>
+        private protected AmazonQLDBSessionConfig SessionConfig { get; set; } = null;
+
+        protected void PrepareForBuild()
```
[the remark this answers is not in the document]
Comment (justing-bq) [reply Amazon.QLDB.Driver/driver/BaseQldbDriverBuilder.cs]: Done.

File: Amazon.QLDB.Driver/result/IAsyncResult.cs (hunk 6 above)
Comment (justing-bq) [reply Amazon.QLDB.Driver/result/IAsyncResult.cs]: Done.

File: Amazon.QLDB.Driver/result/AsyncBufferedResult.cs (hunk 7 above)
Comment (justing-bq) [reply Amazon.QLDB.Driver/result/AsyncBufferedResult.cs:27]: Done.
